### ISSUE TITLE
Activating a specific Python module

### ISSUE THREAD

**Issue by adamjstewart** [LLNL/spack#1178]
I have several Python modules, some installed before the newarch support and some after:

```
$ spack find -v -l py-numpy
==> 6 installed packages.
--  / gcc@4.4.7 -------------------------------------------------
fbmftzz py-numpy@1.10.4+blas+lapack

--  / gcc@5.3.0 -------------------------------------------------
kdkkapz py-numpy@1.10.4+blas+lapack

--  / gcc@6.1.0 -------------------------------------------------
x3736ug py-numpy@1.11.0+blas+lapack  kgfszlb py-numpy@1.11.0+blas+lapack
tduopn7 py-numpy@1.11.0+blas+lapack

-- linux-centos6-x86_64 / gcc@6.1.0 -----------------------------
j3gy5iw py-numpy@1.11.0+blas+lapack
```

How can I activate a specific module? Specifying the arch doesn't work and picks up all versions for some reason. Activate doesn't accept a hash, although uninstall does? Are we using completely different spec parsers for every subcommand?? Anyway, specifying the hash doesn't seem to work even for uninstall:

```
$ spack uninstall py-numpy /j3gy5iw
==> Error: py-numpy matches multiple packages:

--  / gcc@4.4.7 -------------------------------------------------
fbmftzz py-numpy@1.10.4%gcc+blas+lapack


--  / gcc@5.3.0 -------------------------------------------------
kdkkapz py-numpy@1.10.4%gcc+blas+lapack


--  / gcc@6.1.0 -------------------------------------------------
tduopn7 py-numpy@1.11.0%gcc+blas+lapack

x3736ug py-numpy@1.11.0%gcc+blas+lapack

kgfszlb py-numpy@1.11.0%gcc+blas+lapack


-- linux-centos6-x86_64 / gcc@6.1.0 -----------------------------
j3gy5iw py-numpy@1.11.0%gcc+blas+lapack


==> Error: py-numpy@1.11.0%gcc@6.1.0+blas+lapack arch=linux-centos6-x86_64^bzip2@1.0.6%gcc@6.1.0 arch=linux-centos6-x86_64^ncurses@6.0%gcc@6.1.0 arch=linux-centos6-x86_64^openblas@0.2.18%gcc@6.1.0+fpic~openmp+shared arch=linux-centos6-x86_64^openssl@1.0.2h%gcc@6.1.0 arch=linux-centos6-x86_64^py-nose@1.3.7%gcc@6.1.0 arch=linux-centos6-x86_64^py-setuptools@20.7.0%gcc@6.1.0 arch=linux-centos6-x86_64^python@2.7.11%gcc@6.1.0 arch=linux-centos6-x86_64^readline@6.3%gcc@6.1.0 arch=linux-centos6-x86_64^sqlite@3.8.5%gcc@6.1.0 arch=linux-centos6-x86_64^zlib@1.2.8%gcc@6.1.0 arch=linux-centos6-x86_64 matches multiple packages:

--  / gcc@6.1.0 -------------------------------------------------
tduopn7 py-numpy@1.11.0%gcc+blas+lapack


-- linux-centos6-x86_64 / gcc@6.1.0 -----------------------------
j3gy5iw py-numpy@1.11.0%gcc+blas+lapack


==> Error: You can either:
    a) Use a more specific spec, or
    b) use spack uninstall -a to uninstall ALL matching specs.
```

**Comment by tgamblin** [comment]
This is a bug with the way `satisfies()` works with architectures.

**Comment by adamjstewart** [comment]
Note that this also affects `spack load`. I have no way of running `spack load` anymore, and the only way to uninstall a package is to delete it outside of spack and run `spack reindex`.

**Comment by adamjstewart** [comment]
I have another package in a similar situation that I was able to uninstall by running `spack uninstall /iqzw5zo` instead of `spack uninstall armadillo /iqzw5zo`

**Comment by becker33** [comment]
I think part of the issue here is that satisfies does not check hashes. So if we have a syntax change (like the "new" architecture support) such that two packages match for satisfaction but with different hashes, satisfies doesn't properly distinguish them. 

I'll have to look into it, but comparing hashes when calling satisfies on two concrete specs with cached hashes may resolve this.

**Comment by tgamblin** [comment]
@adamjstewart: should be fixed by #2626.  Can you verify?
